pkg/cloud/services/eks/addons.go

```diff
@@ -0,0 +1,188 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eks"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	ekscontrolplanev1 "sigs.k8s.io/cluster-api-provider-aws/controlplane/eks/api/v1alpha3"
+	eksaddons "sigs.k8s.io/cluster-api-provider-aws/pkg/eks/addons"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/record"
+)
+
+func (s *Service) reconcileAddons(ctx context.Context) error {
+	s.scope.Info("Reconciling EKS addons")
+
+	eksClusterName := s.scope.KubernetesClusterName()
+
+	// 1. get addons from AWS for cluster
+	s.scope.V(2).Info("getting installed eks addons", "cluster", eksClusterName)
+	installed, err := s.getClusterAddons()
+	if err != nil {
+		return fmt.Errorf("getting installed eks addons: %w", err)
+	}
+
+	// 2. get addons from spec
+	desiredAddons := s.translateToSdkAddon(s.scope.Addons())
+
+	// 3. Exit early if there is none
+	if len(installed) == 0 && len(desiredAddons) == 0 {
+		s.scope.Info("no addons installed and no addons to install, no action needed")
+		return nil
+	}
+
+	// 3. compute CUD operations between 1 & 2
+	s.scope.V(2).Info("creating eks addons plan", "cluster", eksClusterName, "numdesired", len(desiredAddons), "numinstalled", len(installed))
+	addonsPlan := eksaddons.NewAddonsPlan(eksClusterName, desiredAddons, installed, s.EKSClient)
+	procedures, err := addonsPlan.Create(ctx)
+	if err != nil {
+		s.scope.Error(err, "failed creating eks addons plane")
+		return fmt.Errorf("creating eks addons plan: %w", err)
+	}
+	s.scope.V(2).Info("computed EKS addons plan", "numprocs", len(procedures))
+	if len(procedures) == 0 {
+		s.scope.Info("addon plan has no actions required, no action needed")
+		return nil
+	}
+
+	// 4. Perform required operations
+	for _, procedure := range procedures {
+		s.scope.V(2).Info("Executing addon procedure", "name", procedure.Name())
+		if err := procedure.Do(ctx); err != nil {
+			s.scope.Error(err, "failed executing addon procedure", "name", procedure.Name())
+			return fmt.Errorf("%s: %w", procedure.Name(), err)
+		}
+	}
+
+	// 5. Update status with addons installed details
+	// NOTE: we should be able to construct this froom the plan/procedures????
+	s.scope.V(2).Info("getting installed eks addons for status", "cluster", eksClusterName)
+	installed, err = s.getClusterAddons()
+	if err != nil {
+		return fmt.Errorf("getting installed eks addons: %w", err)
+	}
+	addonStatus := []ekscontrolplanev1.AddonStatus{}
+	for _, installedAddon := range installed {
+		addonStatus = append(addonStatus, ekscontrolplanev1.AddonStatus{
```

Comment:
may have missed it but can we show `Status` and if appropriate `Health` on the resource?

Reply:
I have added status and health/issues